File: Vienna/Sources/Criteria/Criteria+SQL.m
//
//  Criteria+SQL.m
//  Vienna
//
//  Copyright 2022 Tassilo Karge
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  https://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

#import "Criteria+SQL.h"
#import "Database.h"
#import "Field.h"
#import "Article.h"

@implementation CriteriaTree (SQL)

-(NSString *)stringForCriteria:(Criteria *)criteria database:(Database *)database {

    NSMutableString *sqlString = [NSMutableString string];

    Field * field = [database fieldByName:criteria.field];
    NSAssert1(field != nil, @"Criteria field %@ does not have an associated database field", [criteria field]);

    NSString * operatorString = nil;
    NSString * valueString = nil;

    switch (criteria.operatorType) {
        case VNACriteriaOperatorEqualTo:
            if (field.type == VNAFieldTypeString) {
                operatorString = @"='%@'"; break;
            } else {
                operatorString = @"=%@"; break;
            }
        case VNACriteriaOperatorNotEqualTo:
            if (field.type == VNAFieldTypeString) {
                operatorString = @"<>'%@'"; break;
            } else {
                operatorString = @"<>%@"; break;
            }
        case VNACriteriaOperatorLessThan:
            operatorString = @"<%@"; break;
        case VNACriteriaOperatorGreaterThan:
            operatorString = @">%@"; break;
        case VNACriteriaOperatorLessThanOrEqualTo:
            operatorString = @"<=%@"; break;
        case VNACriteriaOperatorGreaterThanOrEqualTo:
            operatorString = @">=%@"; break;
        case VNACriteriaOperatorContains:
            operatorString = @" LIKE '%%%@%%'"; break;
        case VNACriteriaOperatorContainsNot:
            operatorString = @" NOT LIKE '%%%@%%'"; break;
        case VNACriteriaOperatorBefore:
            operatorString = @"<%@"; break;
        case VNACriteriaOperatorAfter:
            operatorString = @">%@"; break;
        case VNACriteriaOperatorOnOrBefore:
            operatorString = @"<=%@"; break;
        case VNACriteriaOperatorOnOrAfter:
            operatorString = @">=%@"; break;

        case VNACriteriaOperatorUnder:
        case VNACriteriaOperatorNotUnder:
            // Handle the operatorString later. For now just make sure we're working with the
            // right field types.
            NSAssert([field type] == VNAFieldTypeFolder, @"Under operators only valid for folder field types");
            break;
    }

    // Unknown operator - skip this clause
    if (operatorString == nil) {
        return nil;
    }

    switch (field.type)
    {
        case VNAFieldTypeFlag:
            valueString = [criteria.value isEqualToString:@"Yes"] ? @"1" : @"0";
            break;
        case VNAFieldTypeFolder: {
            Folder * folder = [database folderFromName:criteria.value];
            NSString *scopeString = [database sqlScopeForFolder:folder criteriaOperator:criteria.operatorType];
            [sqlString appendString:scopeString];
            break;
        }
        case VNAFieldTypeDate: {
            NSCalendar *calendar = NSCalendar.currentCalendar;
            NSDate *startDate = [calendar startOfDayForDate:[NSDate date]];
            NSString * criteriaValue = criteria.value.lowercaseString;
            NSCalendarUnit calendarUnit = NSCalendarUnitDay;

            // "yesterday" is a short hand way of specifying the previous day.
            if ([criteriaValue isEqualToString:@"yesterday"]) {
                startDate = [calendar dateByAddingUnit:NSCalendarUnitDay
                                                 value:-1
                                                toDate:startDate
                                               options:0];
            }
            // "last week" is a short hand way of specifying a range from 7 days ago to today.
            else if ([criteriaValue isEqualToString:@"last week"]) {
                startDate = [calendar dateByAddingUnit:NSCalendarUnitWeekOfYear
                                                 value:-1
                                                toDate:startDate
                                               options:0];
                calendarUnit = NSCalendarUnitWeekOfYear;
            }

            if (criteria.operatorType == VNACriteriaOperatorEqualTo) {
                NSDate *endDate = [calendar dateByAddingUnit:calendarUnit
                                                       value:1
                                                      toDate:startDate
                                                     options:0];
                NSString *dateIs = [NSString stringWithFormat:@"( %@>=%f AND %@<%f )", field.sqlField, startDate.timeIntervalSince1970, field.sqlField, endDate.timeIntervalSince1970];
                [sqlString appendString:dateIs];
            }
            else {
                if ((criteria.operatorType == VNACriteriaOperatorAfter) || (criteria.operatorType == VNACriteriaOperatorOnOrBefore)) {
                    startDate = [calendar dateByAddingUnit:NSCalendarUnitDay
                                                     value:1
                                                    toDate:startDate
                                                   options:0];
                }
                valueString = [NSString stringWithFormat:@"%f", startDate.timeIntervalSince1970];
            }
            break;
        }
        case VNAFieldTypeString:
            if (field.tag == ArticleFieldIDText) {
                // Special case for searching the text field. We always include the title field in the
                // search so the resulting SQL statement becomes:
                //
                //   (text op value or title op value)
                //
                // where op is the appropriate operator.
                //
                Field * titleField = [database fieldByName:MA_Field_Subject];
                NSString * value = [NSString stringWithFormat:operatorString, criteria.value];
                NSString *op = criteria.operatorType == VNACriteriaOperatorNotEqualTo || criteria.operatorType == VNACriteriaOperatorContainsNot ? @"AND" : @"OR";
                [sqlString appendFormat:@"(%1$@%2$@ %3$@ %4$@%2$@)", field.sqlField, value, op, titleField.sqlField];
                break;
            }
        case VNAFieldTypeInteger:
            valueString = [NSString stringWithFormat:@"%@", criteria.value];
            break;
    }

    if (valueString != nil) {
        [sqlString appendString:field.sqlField];
        [sqlString appendFormat:operatorString, valueString];
    }

    return sqlString;
}

/* criteriaToSQLForDatabase
 * Converts a criteria tree to it's SQL representative.
 */
-(NSString *)toSQLForDatabase:(Database *)database
{
    NSMutableString * sqlString = [NSMutableString string];
    NSInteger count = 0;

    for (NSObject<CriteriaElement> *criteria in self.criteriaTree)
    {
        NSString *conditionString = @"";
        if (count++ > 0) {
            switch (self.condition) {
            case VNACriteriaConditionAny:
                conditionString = @" OR ";
                break;
            case VNACriteriaConditionNone:
                conditionString = @" AND NOT ";
                break;
            case VNACriteriaConditionAll:
            default:
                conditionString = @" AND ";
                break; //Unknown condition
            }
        } else if (self.condition == VNACriteriaConditionNone) {
            conditionString = @"NOT ";
        }

        [sqlString appendString:conditionString];

        if ([criteria isKindOfClass:[CriteriaTree class]]) {
            [sqlString appendString:[NSString stringWithFormat:@"( %@ )", [(CriteriaTree *)criteria toSQLForDatabase:database]]];
        } else {
            [sqlString appendString:[self stringForCriteria:(Criteria *)criteria database:database]];
        }
    }
    return sqlString;
}

@end

File: Vienna/Sources/Criteria/Criteria+SQL.swift
//
//  Criteria+SQL.swift
//  Vienna
//
//  Copyright 2023 Tassilo Karge
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  https://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

import Foundation

@objc
protocol SQLConversion {
    @objc(toSQLForDatabase:)
    func toSQL(database: Database) -> String
}

@objc
extension CriteriaTree: SQLConversion {
    func toSQL(database: Database) -> String {

        let sqlOperator: String
        let noneConditionPrefix: String
        switch condition {
        case .any:
            noneConditionPrefix = ""
            sqlOperator = "OR"
        case .none:
            noneConditionPrefix = "NOT "
            sqlOperator = "AND NOT"
        case .invalid, .all:
            noneConditionPrefix = ""
            sqlOperator = "AND"
        }

        return noneConditionPrefix + criteriaTree.map { crit in
            if let crit = crit as? CriteriaTree {
                return "( \(crit.toSQL(database: database)) )"
            } else if let crit = crit as? Criteria {
                return crit.toSQL(database: database)
            } else {
                fatalError("faulty criteria type \(type(of: crit))")
            }
        }.joined(separator: " \(sqlOperator) ")
    }
}

@objc
extension Criteria: SQLConversion {
    func toSQL(database: Database) -> String {
        guard let databaseField = database.field(byName: field), let sqlField = databaseField.sqlField else {
            fatalError("Criteria field \(field) does not have an associated database field")
        }

        let sqlFieldName = "\"\(sqlField)\""
        switch databaseField.type {
        case .flag:
            return flagSqlString(sqlFieldName: sqlFieldName)
        case .date:
            return dateSqlString(sqlFieldName: sqlFieldName)
        case .folder:
            return folderSqlString(sqlFieldName: sqlFieldName, database: database)
        case .integer:
            return integerSqlString(sqlFieldName: sqlFieldName)
        case .string:
            if databaseField.name == MA_Field_Text {
                // Special case for searching the text field: We always include the title field in the search
                // TODO: decide how to migrate this now that we allow nested criteria
                return "(\(stringSqlString(sqlFieldName: sqlFieldName)) OR \(Criteria(field: MA_Field_Subject, operatorType: operatorType, value: value).toSQL(database: database)))"
            } else {
                return stringSqlString(sqlFieldName: sqlFieldName)
            }
        @unknown default:
            fatalError("unimplemented criteria field type \(databaseField.type)")
        }
    }

    func dateSqlString(sqlFieldName: String) -> String {
        let startOfToday = Calendar.current.startOfDay(for: Date())

        let startDate: Date?
        if value == DateOffset.today.rawValue {
            startDate = startOfToday
        } else if value == DateOffset.yesterday.rawValue {
            startDate = Calendar.current.date(byAdding: .day, value: -1, to: startOfToday)
        } else if value == DateOffset.lastWeek.rawValue {
            startDate = Calendar.current.date(byAdding: .weekOfYear, value: -1, to: startOfToday)
        } else {
            // Check for the pattern for date with unit criteria
            let datePatternRegex = try? NSRegularExpression(pattern: "^([0-9]+) (.+)$")
            let datePatternMatch = datePatternRegex?.firstMatch(in: value, range: NSRange(location: 0, length: value.utf16.count))
            if let datePatternMatch = datePatternMatch {
                guard let valueRange = Range(datePatternMatch.range(at: 1), in: value),
                      let amount = Int(value[valueRange]),
                      let unitRange = Range(datePatternMatch.range(at: 2), in: value),
                      let unit = DateUnit(rawValue: String(value[unitRange]))
                else {
                      fatalError("Malformed date predicate value: \(value)")
                }

                startDate = Calendar.current.date(byAdding: unit.calendarComponent, value: -amount, to: Date())
            } else {
                fatalError("Unknown value \(value) for date field")
            }
        }

        let startOfDayToEndOfDay = DateComponents(calendar: Calendar.current, day: 1, second: -1)
        guard let startOfDay = startDate, let endOfDay = Calendar.current.date(byAdding: startOfDayToEndOfDay, to: startOfDay) else {
            fatalError("Start of day or end of day date calculation for \(value) failed")
        }

        let endOfDayInterval = Int64(endOfDay.timeIntervalSince1970)
        let startOfDayInterval = Int64(startOfDay.timeIntervalSince1970)

        switch operatorType {
        case .equalTo:
            return "\(sqlFieldName) >= \(startOfDayInterval) AND \(sqlFieldName) <= \(endOfDayInterval)"
        case .before:
            return "\(sqlFieldName) < \(startOfDayInterval)"
        case .after:
            return "\(sqlFieldName) > \(endOfDayInterval)"
        case .onOrBefore:
            return "\(sqlFieldName) <= \(endOfDayInterval)"
        case .onOrAfter:
            return "\(sqlFieldName) >= \(startOfDayInterval)"
        default:
            fatalError("Illegal operator \(operatorType) for date field")
        }
    }

    func flagSqlString(sqlFieldName: String) -> String {
        guard operatorType == .equalTo || operatorType == .notEqualTo else {
            fatalError("Operator type \(operatorType) not applicable to flag field \(sqlFieldName)")
        }
        let sqlValue: String
        if value == "Yes" { sqlValue = "1" } else { sqlValue = "0" }
        let sqlOperator = standardSqlOperator()
        return "\(sqlFieldName) \(sqlOperator) \(sqlValue)"
    }

    func folderSqlString(sqlFieldName: String, database: Database) -> String {
        // trim extra spaces added by predicate editor's formatting
        let cleanedValue = String(value.drop { $0 == " " })
        let folder = database.folder(fromName: cleanedValue)
        let scope: QueryScope
        switch operatorType {
        case .under:
            scope = [.subFolders, .inclusive]
        case .notUnder:
            scope = .subFolders
        case .equalTo:
            scope = .inclusive
        case .notEqualTo:
            scope = []
        default:
            fatalError("Operator type \(operatorType) not applicable to folder field \(sqlFieldName)")
        }
        return database.sqlScope(for: folder, flags: scope, field: sqlFieldName)
    }

    func integerSqlString(sqlFieldName: String) -> String {
        return "\(sqlFieldName) \(standardSqlOperator()) \(value)"
    }

    func stringSqlString(sqlFieldName: String) -> String {
        let sqlOperator: String
        switch operatorType {
        case .contains:
            sqlOperator = "LIKE '%%%@%%'"
        case .containsNot:
            sqlOperator = "NOT LIKE '%%%@%%'"
        default:
            sqlOperator = "\(standardSqlOperator()) '%@'"
        }
        let formattedOperator = String(format: sqlOperator, value)
        return "\(sqlFieldName) \(formattedOperator)"
    }

    func standardSqlOperator() -> String {
        switch operatorType {
        case .equalTo:
            return "="
        case .notEqualTo:
            return "<>"
        case .lessThan:
            return "<"
        case .greaterThan:
            return ">"
        case .lessThanOrEqualTo:
            return "<="
        case .greaterThanOrEqualTo:
            return ">="
        default:
            fatalError("No standard sql operator for type \(operatorType)")
        }
    }

}
